src/api/server.rs

```diff
@@ -0,0 +1,123 @@
+use super::{handler, schema::build_schema};
+
+use async_graphql::{
+    http::{playground_source, GraphQLPlaygroundConfig},
+    QueryBuilder,
+};
+use async_graphql_warp::{graphql_subscription, GQLResponse};
+use std::convert::Infallible;
+use std::net::SocketAddr;
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::{Receiver, Sender};
+use warp::filters::BoxedFilter;
+use warp::{http::Response, Filter, Reply};
+
+pub struct Server {
+    /// Address for the API server to bind on
+    address: SocketAddr,
+
+    /// Transmission channel to trigger closure of a running API server
+    trigger_cancel: Sender<()>,
+
+    /// Receiver signal to cancel running API server
+    cancel_signal: Option<Receiver<()>>,
+}
+
+impl Server {
+    /// Returns a new API Server
+    pub fn new(address: SocketAddr) -> Server {
+        let (trigger_cancel, cancel_signal) = oneshot::channel::<()>();
+
+        Server {
+            address,
+            trigger_cancel,
+            cancel_signal: Some(cancel_signal),
+        }
+    }
+
+    /// String representation of the bound IP address
+    pub fn ip(&self) -> String {
+        self.address.ip().to_string()
+    }
+
+    /// String representation of the bound port
+    pub fn port(&self) -> String {
```

Comment:
Might make sense to just return `u16` here, and have the callee convert the port to a string if need be.

Reply:
Thanks. I did this because the only current usage is logging, which is otherwise `port().to_string().as_str()`, which felt a bit aesthetically unpleasing. I think it's probably a better API to return the actual thing and let the callee transform it though, I think.

Reply:
Sounds good to me 👍

> which is otherwise `port().to_string().as_str()`, which felt a bit aesthetically unpleasing

You should be able to dereference (`&`) the string, since `String` implements the `Deref` trait, which returns a `&str`. This makes it a bit more pleasant to read.

By the way, one more convention I learned a while back when digging into Rust is to use `to_owned()` instead of `to_string()` to go from a `&str` to a `String`.

Someone (somewhere) explained it as being more natural, as it feels a bit strange that you already have a string (`&str`) on which you call `to_string()`. `to_owned()` makes it very explicit that you are about to allocate memory to get an owned copy of the string.

This doesn't always work though, specifically in this case, you're going from `IpAddr` to `String`, which requires you to use the `ToString` trait, as `ToOwned` simply returns an owned `IpAddr`.